opal/mca/btl/openib/btl_openib_component.c

```diff
@@ -227,6 +227,82 @@ static int btl_openib_component_close(void)
     return rc;
 }
 
+/*
+ * Filter on device type.  Here we're weeding out IB devices by
+ * default.  Starting with Open MPI 4.0 we don't support Open IB for
+ * IB HCAs unless the user specifically requested to override this
+ * policy.
+ *
+ * Return true if device allowed owing to above policy, otherwise false
+ */
+
+static bool allow_device(struct ibv_device *ib_dev)
+{
+    bool ret = true;
+
+#if defined(HAVE_STRUCT_IBV_DEVICE_TRANSPORT_TYPE)
+#if HAVE_DECL_IBV_LINK_LAYER_ETHERNET
+    int i;
+    struct ibv_context *dev_context = NULL;
+    struct ibv_port_attr ib_port_attr;
+    struct ibv_device_attr device_attr;
+#endif
+
+    if (false == mca_btl_openib_component.allow_ib) {
+
+        /*
+         * first if its an iWarp device we've found its easy
+         */
+        if (IBV_TRANSPORT_IWARP == ib_dev->transport_type) {
+            goto fn_exit;
+        }
+
+        /*
+         * well it may , or it may not be a RoCE device, so
+         * do some extra stuff.
+         *
+         * if we encounter a problem, or no active ports, return
+         * true as the device will be disallowed later anyway.
+         */
+
+#if HAVE_DECL_IBV_LINK_LAYER_ETHERNET
+
+        dev_context = ibv_open_device(ib_dev);
+        if (NULL == dev_context) {
+            goto fn_exit;
+        }
+
+        if (0 != ibv_query_device(dev_context, &device_attr)) {
+            goto fn_exit;
+        }
+
+        for (i = 1; i <= device_attr.phys_port_cnt; i++) {
+             if (0 != ibv_query_port(dev_context, i, &ib_port_attr)) {
+                 goto fn_exit;
+             }
+
+            if ((IBV_PORT_ACTIVE == ib_port_attr.state) &&
+                 (IBV_LINK_LAYER_INFINIBAND == ib_port_attr.link_layer)) {
+                ret = false;
+                goto fn_exit;
```

Comment:
I don't think that this is quite right.

I think you want to reject the device if *all* the ports are IB.  If any of them are Ethernet, then you can allow the overall *device*, but you'll need to reject the individual IB *ports* later (perhaps in `init_one_port()`?).